### DISCUSSION TITLE
FLAML returns XGB model with max_depth=0

### DISCUSSION THREAD

**Discussion by jaukiuki** [microsoft/FLAML#211] [Q&A]
I try to use FLAML in my modelling project I use this code:

_automl_clf = AutoML()
estimators = ['xgboost']
automl_clf.fit(X_train, y_train, task="classification", time_budget=1800, eval_method='CV', n_splits=5, n_jobs=96,
               estimator_list=estimators, max_iter=5000, X_val=X_test, y_val=y_test)_

I've reveiced information in my log as below, please look at parameter **max_depth=0**. How it's possible? In this situation a cannot use this in real xgboost code and check returned parameters?

Can anyone explain to me what is going on here?

_[flaml.automl: 09-16 16:04:29] {2021} INFO - selected model: XGBClassifier(colsample_bylevel=0.4173380937708115,
              colsample_bytree=0.6342278976506369, grow_policy='lossguide',
              learning_rate=0.05598416814070667, max_depth=0, max_leaves=369,
              min_child_weight=0.3156717050599599, n_estimators=1058, n_jobs=96,
              reg_alpha=0.026223717746293316, reg_lambda=0.7095314560287201,
              subsample=0.8086890874802936, tree_method='hist',
              use_label_encoder=False, verbosity=0)
[flaml.automl: 09-16 16:04:29] {1529} INFO - fit succeeded_

**Comment by sonichi** [comment]
`max_depth=0` means no restriction on the depth. `max_leaves` and `min_child_weight` govern the tree size. It is real xgboost.

**Reply by jaukiuki** [reply]
Thank you for answer.
I see that  max_depth=0 is only accepted when grow_policy='lossguide' and tree_method is set as hist or gpu_hist but does flaml can use grow_policy=’ depthwise’ and tree_method=’exact’?

**Reply by sonichi** [reply]
The current grow_policy is set to 'lossguide' and tree_method is 'hist'. 
If you'd like to change `grow_policy` and `tree_method`, you can define a custom learner by inheriting from XGBoostSklearnEstimator, and override its constructor, in a similar way as in Section 4 of this [notebook](https://render.githubusercontent.com/view/ipynb?color_mode=auto&commit=b1115d5347e00326d10e5d20836b912bb2cc95e9&enc_url=68747470733a2f2f7261772e67697468756275736572636f6e74656e742e636f6d2f6d6963726f736f66742f464c414d4c2f623131313564353334376530303332366431306535643230383336623931326262326363393565392f6e6f7465626f6f6b2f666c616d6c5f7867626f6f73742e6970796e62&nwo=microsoft%2FFLAML&path=notebook%2Fflaml_xgboost.ipynb&repository_id=289107293&repository_type=Repository#4.-Add-customized-XGBoost-learners-in-FLAML).
